### PR TITLE
Add a glossary.

### LINKED ISSUE #3380: FAQ docs request: What is the distinction between 'crate' and 'package'?
I spoke with @steveklabnik a couple nights ago and they informed that the terms 'crate' and 'package' are different concepts in the context of cargo. Someone in the #rust-tools channel on IRC just asked what the differences are. It might be worth clarifying this in the [Frequently Asked Questions section](http://doc.crates.io/faq.html) (or maybe a different section?).

I'd be happy to write-up something for this, but I actually don't recall the distinctions from the conversation :)

### LINKED ISSUE #4392: [book] Add a Glossary section?
What terms are used in Cargo and what they mean is a common theme in Cargo issues/pulls, as well as in the documentation. Recently I just updated some of the docs to use the correct term in their place. (See https://github.com/rust-lang/cargo/pull/4376 and https://github.com/rust-lang/cargo/pull/4375)

I think with the expansion of Cargo features, it makes sense to have a Glossary document that describes each term and points to the detailed description.

We can add this to the new mdBook-based docs, and not in-repo.

Here's some example terms that I have seen confusions about in the docs and public spaces, like the users forum:
* crate
* package
* build target (as in platform and its binary format)
* target (as in lib, bins, tests, examples, ...)
* crate file
* package repository (as in git repo)
* package registry (as in crates.io), a.k.a. package index
* package root (the directory)
* workspace root (the directory)
* workspace root package (a package or its manifest)
* repository root (the directory)

What do you think?

### PR DESCRIPTION
**Pull request by ehuss**
Closes #4392, closes #3380.

### CODE REVIEW

File: src/doc/src/appendix/glossary.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+A [*feature*][feature] is a named flag which allows for conditional
+compilation. A feature can refer to an optional dependency, or an arbitrary
+name defined in a `Cargo.toml` manifest that can be checked within source
+code.
+
+Cargo has [*unstable feature flags*][cargo-unstable] which can be used to
+enable experimental behavior of Cargo itself. The Rust compiler also has its
+own unstable feature flags (see [The Unstable Book][unstable-book]) and
+Rustdoc also has its own unstable feature flags (see [The Rustdoc
+Book][rustdoc-unstable]).
```
Comment (QuietMisdreavus) [inline src/doc/src/appendix/glossary.md]: I feel like the last sentence here could be phrased better as:

> The Rust compiler and Rustdoc also have their own unstable feature flags (see [The Unstable Book] and [The Rustdoc Book]).

File: src/doc/src/appendix/glossary.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+### Spec
+
+See [package id specification](#package).
+
+### Target
+
+The meaning of the term *target* depends on the context:
+
+- **Target Crate** — Cargo packages consist of *targets* which correspond to
```
[the remark this answers is not in the document]
Comment (ehuss) [reply src/doc/src/appendix/glossary.md]: Hm, yea, both seem kinda odd.  How about "Cargo Target"?  That seems a little less vague.

[the remark this answers is not in the document]
Comment (dwijnand) [reply src/doc/src/appendix/glossary.md]: Sure, let's go with that!

**Pull request merged**